keras_nlp/layers/cached_multi_head_attention.py

```diff
@@ -0,0 +1,116 @@
+
+    In autoregressive decoding, it's a common practice to cache the key/value in
+    multi-head attention of previously seen tokens in order to make the
+    computation faster. With cached K and V, we can compute the attention output
+    of the last token without recomputing the forward pass for previously seen
+    tokens. This caching method is only useful during decoding, and should not
+    be used during training.
+
+    Call arguments:
+        query: Query `Tensor` of shape `(B, T, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`.
+        value: Value `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`.
+        key: Optional key `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`. If not given, will use `value` for both
+            `key` and `value`, which is the most common case.
+        attention_mask: a boolean mask of shape `(B, T, S)` if `cache=None`,
+            otherwise `(B, 1, S)`. `attention_mask` prevents
+            attention to certain positions. The boolean mask specifies which
+            query elements can attend to which key elements, 1 indicates
+            attention and 0 indicates no attention. Broadcasting can happen for
+            the missing batch dimensions and the head dimension.
+        cache: a dense float Tensor. The cache of key/value of leading tokens.
+            `cache` is of shape [2, B, max_seq_len, num_heads, key_dims].
+        cache_index: a int or int Tensor, the index of the current token being
+            processed. If `cache_index=None` while `cache` is set, it means
+            it's the first pass to build the cache.
+
+    Returns:
+        attention_output: The result of the computation, of shape `(B, T, E)`,
+            where `T` is for target sequence shapes and `E` is the query input
+            last dimension if `output_shape` is `None`. Otherwise, the
+            multi-head outputs are projected to the shape specified by
+            `output_shape`.
+        cache: the updated cache.
+    """
+
+    def call(
+        self,
+        query,
+        value,
+        key=None,
+        attention_mask=None,
+        use_causal_mask=True,
+        cache=None,
+        cache_index=None,
+    ):
+        if not self._built_from_signature:
+            self._build_from_signature(query=query, value=value, key=key)
+        if key is None:
+            key = value
+
+        attention_mask = self._compute_attention_mask(
+            query,
+            value,
+            key=key,
+            attention_mask=attention_mask,
+            use_causal_mask=use_causal_mask,
+        )
+
+        query = self._query_dense(query)
+        key = self._key_dense(key)
+        value = self._value_dense(value)
+
+        if cache is not None:
+            if cache_index is None:
+                seq_len = tf.shape(query)[1]
+                k_update_indices = [0, 0, 0, 0, 0]
```

Comment:
We could use the unstack stack trick here. Then you can have the same `update_indices` for each dynamic_update_slice.

Reply:
simply doing stack to unify the `update_indices ` for k and v isn't very helpful to me.

Reply:
Let me share some code. May make sense to cover this in a follow up.

Reply:
Here's some code showing where we could head for a simplified implementation of `CachedMultiHeadAttention`. https://github.com/mattdangerw/keras-nlp/commit/a4a9e4ba3c1e5849b07defc91c55a848cec9d73e. Let's discuss offline, but probably we can just do this as a follow up.

Reply:
Talked offline. Let's just do this as a follow up. Getting the mask right is a bit tricky, so we can leave that for later.